### DISCUSSION TITLE
Weird behavior when declaring route in an entity

### DISCUSSION THREAD

**Discussion by JoDrm** [api-platform/api-platform#2398] [Q&A]
I realized after days of bugging and researching that the order of the routes (new Get,new Post...) matters.
For example :

If I put in this order :
```
new Get(
  uriTemplate: '/users/check/me',
  name: 'api_users_me',
),
new Get(
  normalizationContext: [
      "groups" => [self::ME, User::READ, Licence::READ, Module::READ, DefaultEntityInterface::READ]
  ],
  security: "is_granted('" . RightsConst::USER_READ . "', object)",
),
```

It's return me user object with that : 
```
 "userTeams": [
        {
            "id": 1,
            "user": "/users/check/me", <-- WRONG IRI
            "team": "/teams/3",
            "roles": [],
            "rights": [],
            "iri": "/user-teams/1",
            "iriBase": "/user-teams/"
        }
    ],
```

If I put in this order : 

```
new Get(
    normalizationContext: [
        "groups" => [self::ME, User::READ, Licence::READ, Module::READ, DefaultEntityInterface::READ]
    ],
    security: "is_granted('" . RightsConst::USER_READ . "', object)",
),

new Get(
    uriTemplate: '/users/check/me',
    name: 'api_users_me',
),
```

It's return me user object with that :

```
   "userTeams": [
        {
            "id": 1,
            "user": "/users/1", <-- GOOD IRI
            "team": "/teams/3",
            "roles": [],
            "rights": [],
            "iri": "/user-teams/1",
            "iriBase": "/user-teams/"
        }
    ],
```

This also causes other problems like the error "unable to generate IRI ...."

Is this the normal expected behavior?

And if it is why it's not explicitly write in the api-platform documentation ? 

IF THE DECLARATION POSITION OF A ROUTE IN THE ENTITY IS IMPORTANT IT SHOULD BE SAYED, MOREOVER THIS BEHAVIOR IS TOTALLY ABBERANT IN MY OPINION.

**Comment by soyuka** [comment]
Operations have a `priority` and the order matters to find the correct IRI. There's also an option to use the `itemUriTemplate` within collection if you want to pick another operation on resolution see: https://api-platform.com/docs/core/operations/#defining-which-operation-to-use-to-generate-the-iri

**Reply by JoDrm** [reply]
"itemUriTemplate option only available on GetCollection and Post operations", so I can't do this on my simple route new Get() /users/me ...

**Reply by soyuka** [reply]
No because an API is supposed to be stateless. You can trick that by decorating the IriConverter.

**Comment by JoDrm** [comment]
@theofidry

**Reply by theofidry** [reply]
If I understood your problem correctly you have a user entity and you're registering an extra get method to it and this is causing problems.

The root of the issue is that fundamentally, the extra get method you are adding is not RESTful which results in some frictions with API-Platform. This is why it is not a "happy path" and you're having some issues.

They should however, be circumventable:

- either change the declaration order or play with the priorities (I think the priority is clearer, with a comment in the code as to why)
- use a custom IriConverter (you should be able to decorated the API-Platform one without too much trouble, we did that for several cases in our own private apps)

>IF THE DECLARATION POSITION OF A ROUTE IN THE ENTITY IS IMPORTANT IT SHOULD BE SAYED, MOREOVER THIS BEHAVIOR IS TOTALLY ABBERANT IN MY OPINION.

Regarding that point, it actually is obvious it matters. There is two GET routes for a given resource detail, the only two possible behaviours are either to pick the first one (which should be the first one declared, unless corrected via a priority) or _throw_ an error (which in this case would be quite ugly, although maybe better if throw with more context).

It could be documented, but I also suppose this is already and edge-case since non RESTfull which is likely why it was missed. A PR is welcome for that though :)
